### ISSUE TITLE
Add sniff to check best practices around include/require statements to Extra ?

### ISSUE THREAD

**Issue by jrfnl** [WordPress-Coding-Standards/WordPress-Coding-Standards#1143]
There is a sniff available upstream to check that non-conditional file includes use `require(_once)` and that conditional file includes use `include(_once)`. It also checks that no parenthesis are used for these statements (not needed as these are language constructs, not function calls).

I realize this may seem a little opinionated, but this is a valid best practice.

If a `require` fails, it will produce a fatal `E_COMPILE_ERROR` which for an unconditional file include is warranted.
If a `include` fails, it will produce a warning for the file include and possibly a plenitude of other warnings/errors if the include was not used conditionally.

I propose adding this sniff to the `Extra` ruleset.

@WordPress-Coding-Standards/wpcs-collaborators  Opinions ?

Refs:
* http://php.net/manual/en/function.require.php
* http://php.net/manual/en/function.include.php

**Comment by GaryJones** [comment]
> non-conditional file includes use require(_once) and that conditional file includes use include(_once)

First time I've heard of that use case logic, but seems fine enough.



> checks that no parenthesis are used for these statements

+1.

> I propose adding this sniff to the Extra ruleset.

+1.

Have you got a link to the unit tests for the conditional / un-conditional use of include/require please?

**Comment by jrfnl** [comment]
> Have you got a link to the unit tests for the conditional / un-conditional use of include/require please?

Of course ;-)
https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/PEAR/Tests/Files/IncludingFileUnitTest.inc

Oh and before I forget:
* The sniff actually contains an auto-fixer as well (and unit tests for this).
* I tested the unit test case file against WPCS and adding the sniff does not create any message duplication

Oh and this issue is related to #1007 which mentioned the same upstream sniff.

**Comment by JDGrimes** [comment]
It might be worth noting that WordPress standardized on using parentheses for `include` in https://core.trac.wordpress.org/changeset/28479, because in Hack the parentheses are required.

I don't really have a strong opinion on this, although I do tend to use parentheses when the file name is not just a simple string, because it seems more readable at times:

```php
require_once( SOME_PATH . '/some_file.php' );
```

Other times, I omit them though, and I think I've probably moved more toward omitting them over time. 

So I'd be fine with adding that rule to `Extra`.

I don't understand the reasoning as to why `include` should be used here:

```php
if ($test) {
    include_once 'blank.inc';
    include 'blank.inc';
}
```

It makes sense here:

```php
if (include_once 'blank.inc') {
    $var = include_once 'blank.inc';
    if ($var === true) {
    }
}
```

But in the former case I'd use `require`. The file is needed, so the fact that it is only needed under particular conditions doesn't seem relevant, when we've already determined that those conditions are true.

**Comment by jrfnl** [comment]
> It might be worth noting that WordPress standardized on using parentheses for include in https://core.trac.wordpress.org/changeset/28479, because in Hack the parentheses are required.

Though explicit testing/support for HHVM has been removed since: https://core.trac.wordpress.org/ticket/40548

> But in the former case I'd use require. The file is needed, so the fact that it is only needed under particular conditions doesn't seem relevant, when we've already determined that those conditions are true.

Having run some more tests with the sniff, I came to a similar conclusion.

There are five error codes in the sniff - all of which throw errors:
* BracketsNotRequired
* UseIncludeOnce
* UseInclude
* UseRequireOnce
* UseRequire

What about this adding the sniff to `Extra` like this:
* BracketsNotRequired - include, but downgrade to warning
* ~~UseIncludeOnce~~ - exclude it from the ruleset
* ~~UseInclude~~ - exclude it from the ruleset
* UseRequireOnce - include, but downgrade to warning
* UseRequire - include, but downgrade to warning

**Comment by JDGrimes** [comment]
> What about adding the sniff to Extra like this

+1

So basically, we'd warn for using parentheses, and we'd warn when using `include*` instead of `require*` (except in conditionals, where the sniff would usually expect `include`), but we would never warn about using `require*` instead of `include*`.

**Comment by jrfnl** [comment]
@JDGrimes Exactly!
